packages/backend/src/indexers/DiscoveryIndexer.test.ts

```diff
@@ -1,138 +1,265 @@
+          subscribe: () => {},
+        }),
+      )
+
+      expect(await discoveryIndexer.update(1000, 2000)).toEqual(1000)
+      expect(discoveryEngine.discover).toHaveBeenCalledTimes(2)
+      expect(discoveryEngine.discover).toHaveBeenNthCalledWith(
         1,
+        config,
+        BLOCKS[0].blockNumber,
       )
+      expect(discoveryEngine.discover).toHaveBeenNthCalledWith(
+        2,
+        config,
+        BLOCKS[1].blockNumber,
+      )
+      expect(discoveryRepo.add).toHaveBeenCalledTimes(2)
     })
   })
 
   describe(DiscoveryIndexer.prototype.invalidate.name, () => {
-    it('should not run discovery on the targetHeight', async () => {
+    it('should delete the records after targetHeight', async () => {
       const chainId = ChainId.ETHEREUM
       const config = mockConfig()
       const discoveryEngine = mockObject<DiscoveryEngine>({
         discover: async () => [],
       })
       const discoveryRepository = mockObject<DiscoveryRepository>({
-        addOrUpdate: async () => true,
+        deleteAfter: async () => 1,
       })
-      const BLOCK_NUMBER = 10
+      const BLOCK_10 = mockBlock({ blockNumber: 10 })
 
       const discoveryIndexer = new DiscoveryIndexer(
         discoveryEngine,
         config,
         mockObject<BlockNumberRepository>({
-          findAtOrBefore: async () => mockBlock(BLOCK_NUMBER),
+          findAtOrBefore: async () => BLOCK_10,
         }),
+        mockObject<EventRepository>(),
         discoveryRepository,
         mockObject<IndexerStateRepository>(),
         chainId,
         Logger.SILENT,
         mockObject<CacheInvalidationIndexer>({
           subscribe: () => {},
         }),
+        mockObject<EventIndexer>({
+          subscribe: () => {},
+        }),
       )
 
-      expect(await discoveryIndexer.invalidate(1000)).toEqual(1000)
+      expect(await discoveryIndexer.invalidate(1000)).toEqual(
+        BLOCK_10.timestamp.toNumber(),
+      )
       expect(discoveryEngine.discover).toHaveBeenCalledTimes(0)
+      expect(discoveryRepository.deleteAfter).toHaveBeenCalledTimes(1)
+      expect(discoveryRepository.deleteAfter).toHaveBeenNthCalledWith(
+        1,
+        BLOCK_10.blockNumber,
+        chainId,
+      )
+    })
+
+    it('should run on empty repository', async () => {
+      const discoveryIndexer = new DiscoveryIndexer(
+        mockObject<DiscoveryEngine>(),
+        mockConfig(),
+        mockObject<BlockNumberRepository>({
+          findAtOrBefore: async () => undefined,
+        }),
+        mockObject<EventRepository>(),
+        mockObject<DiscoveryRepository>({
+          deleteAfter: async () => 1,
+        }),
+        mockObject<IndexerStateRepository>(),
+        ChainId.ETHEREUM,
+        Logger.SILENT,
+        mockObject<CacheInvalidationIndexer>({
+          subscribe: () => {},
+        }),
+        mockObject<EventIndexer>({
+          subscribe: () => {},
+        }),
+      )
+
+      await discoveryIndexer.invalidate(0)
```

Comment:
Maybe we should add at least some basic assertions here i.e calling does not reject